x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/In.java

```diff
@@ -84,17 +80,21 @@ public boolean foldable() {
 
     @Override
     public Boolean fold() {
+        // Optimization for early return and Query folding to LocalExec
         if (value.dataType() == DataType.NULL) {
             return null;
         }
         if (list.size() == 1 && list.get(0).dataType() == DataType.NULL) {
-            return false;
+            return null;
         }
 
-        Object foldedLeftValue = value.fold();
+        return doFold(value.fold(), Foldables.valuesOf(list, value.dataType()));
+    }
+
+    public static Boolean doFold(Object value, List<Object> values) {
         Boolean result = false;
-        for (Expression rightValue : list) {
-            Boolean compResult = Comparisons.eq(foldedLeftValue, rightValue.fold());
+        for (Object v : values) {
+            Boolean compResult = Comparisons.eq(value, v);
```

Comment:
Let's be explicit about boxing. `Boolean result = Boolean.FALSE`. `if (compResult == Boolean.TRUE) or `compResult.boolean()` ...